src/main/java/threads/task/Main.java

```diff
@@ -0,0 +1,171 @@
+                                        StringBuilder::append,
+                                        sb -> {
+                                            var messages = sb.toString();
+                                            if (messages.endsWith(", "))
+                                                return messages.substring(0, messages.length() - 2);
+                                            return messages;
+                                        }
+                                ))
+                ));
+    }
+
+    private void printMessagesOfInterruptedThreads() {
+        System.out.println("From: " + getCurrentThreadName() + " message: " + messages
+                .get(getCurrentThreadName())
+                .stream()
+                .collect(Collector.of(
+                        StringBuilder::new,
+                        (sb, message) -> sb.append(message).append(", "),
+                        StringBuilder::append,
+                        sb -> {
+                            var messages = sb.toString();
+                            if (messages.endsWith(", "))
+                                return messages.substring(0, messages.length() - 2);
+                            return messages;
+                        }
+                )) + " interrupt ______________________________"
+        );
+    }
+
+    public void isThreadActive() {
+        var isOneOfThreadWroteMoreThan3Messages = getMessagesFromOtherThreads()
+                .values()
+                .stream()
+                .anyMatch(x -> x.size() >= 3);
+        var isCurrentThreadWroteMoreThan3Messages = messages.get(getCurrentThreadName()).size() >= 3;
+        if (isOneOfThreadWroteMoreThan3Messages || isCurrentThreadWroteMoreThan3Messages) {
+            printMessagesOfInterruptedThreads();
+            clearOwnHistory();
+            currentThread().interrupt();
+        }
+    }
+
+    public int getRandomTimeout() {
+        return RANDOM.nextInt(2000) + 1;
+    }
+
+    private void clearHistory() {
+        getMessagesFromOtherThreads().forEach((key, value) -> value.clear());
+    }
+
+    private void clearOwnHistory() {
+        messages.remove(getCurrentThreadName());
+    }
+
+    @SneakyThrows
+    @Override
+    public void run() {
+        while (!currentThread().isInterrupted()) {
+            sleep(getRandomTimeout());
+            synchronized (RANDOM) {
+                if (!messages.isEmpty()) readNewMessages();
+                writeRandomMessage();
+                isThreadActive();
+            }
+        }
+    }
+}
+
+public class Main {
+    @SneakyThrows
+    public static void main(String[] args) {
+        var a = new Thread(new Client(), "a");
+        var b = new Thread(new Client(), "b");
+        var c = new Thread(new Client(), "c");
+        a.start();
+        b.start();
+        c.start();
+        var isAllThreadsAreInterrupted = false;
+        while (!isAllThreadsAreInterrupted) {
```

Comment:
I think we discussed that a simple join to the main thread will be enough to make the main thread wait for others.